executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,430 @@
+        // Permutations have a RHS multiplicity that is equal to the RHS latch.
+        AnalyzedIdentity::Permutation(PermutationIdentity {
+            id, left, right, ..
+        })
+        | AnalyzedIdentity::PhantomPermutation(PhantomPermutationIdentity {
+            id,
+            left,
+            right,
+            ..
+        }) => bus_interaction_pair(*id, id_counter, left, right, Some(right.selector.clone())),
+        // Native lookups do not have a multiplicity.
+        AnalyzedIdentity::Lookup(LookupIdentity {
+            id, left, right, ..
+        }) => bus_interaction_pair(*id, id_counter, left, right, None),
+        AnalyzedIdentity::PhantomLookup(PhantomLookupIdentity {
+            id,
+            left,
+            right,
+            multiplicity,
+            ..
+        }) => bus_interaction_pair(*id, id_counter, left, right, Some(multiplicity.clone())),
+    }
+}
+
+fn bus_interaction_pair<T: FieldElement>(
+    id: u64,
+    id_counter: &mut RangeFrom<u64>,
+    left: &SelectedExpressions<T>,
+    right: &SelectedExpressions<T>,
+    rhs_multiplicity: Option<AlgebraicExpression<T>>,
+) -> Vec<Identity<T>> {
+    // +1 because we want to be sure it is non-zero
+    let interaction_id: T = (id + 1).into();
+    vec![
+        Identity::BusSend(BusSend {
+            id,
+            interaction_id: AlgebraicExpression::Number(interaction_id),
+            selected_tuple: left.clone(),
+        }),
+        Identity::BusReceive(BusReceive {
+            id: id_counter.next().unwrap(),
+            multiplicity: rhs_multiplicity,
+            interaction_id,
+            selected_tuple: right.clone(),
+        }),
+    ]
+}
+
+#[cfg(test)]
+mod test {
+    use powdr_number::GoldilocksField;
+
+    use crate::witgen::data_structures::identity::Identity;
+
+    use super::convert;
+
+    fn assert_correct_bus_interaction_pair(
+        constraint: &str,
+        expected_send: &str,
+        expected_receive: &str,
+        expected_is_receive_unconstrained: bool,
+    ) {
+        let pil_source = format!(
+            r"
+namespace main(4);
+    col fixed right_latch = [0, 1]*;
+    col witness right_selector, left_latch, a, b, multiplicities;
+    {constraint}
+    
+    // Selectors should be binary
+    (left_latch - 1) * left_latch = 0;
+    (right_selector - 1) * right_selector = 0;
+
+    // Some other bus interactions to potentially confuse the matching
+    right_latch $ [a] in right_latch $ [a];
+    left_latch $ [b] in left_latch $ [b];
+"
+        );
+        let analyzed = powdr_pil_analyzer::analyze_string::<GoldilocksField>(&pil_source).unwrap();
+        let identities = convert(&analyzed.identities);
+        let receives = identities
+            .iter()
+            .filter_map(|id| match id {
+                Identity::BusReceive(r) => Some((r.interaction_id, r.clone())),
+                _ => None,
+            })
+            .collect();
+
+        match (&identities[0], &identities[1]) {
```

Comment:
How can we be sure that the ones we are interented in are at the beginning?

Reply:
Because in `analyzed.identities`, they appear in source order, no? And `convert` doesn't change the order.